File: mib_v2_3_2/distrib.py
import random
from itertools import product
from mib_v2_3_2.var import Var

class Distrib:
    """ Clase para el manejo de distibuciones marginales.
    Atributos:
        table (dict): Dicciónario de las probabilidades (tuple: float).
        vars (tuple): Tupla con el orden de las variables del diccionario.
        parents (tuple (optional)): Tupla con el orden de las variables padres del diccionario.
        
    Distrib(tuple,tuple,dict) -> nuevo objeto Distrib
    
    Distrib(tuple,dict) -> nuevo objeto Distrib
    """

    def __init__(self, table:dict, vars:tuple, parents:tuple = None) -> None:
        self.table = table
        self.vars = vars
        self.parents = parents
    
    def P(self) -> float:
        if not self.parents:
            return self._jointP()
        else:
            return self._condP()
    
    def _jointP(self) -> float:
        key = [v.event for v in self.vars]
        return self.table[tuple(key)]
    
    def _condP(self) -> float:
        vars_key = [v.event for v in self.vars]
        indep_key = [v.event for v in self.parents]
        return self.table[tuple(indep_key)][tuple(vars_key)]
    
    def setSample(self) -> None:
        if self.parents:
            indep_key = tuple([v.event for v in self.parents])
            values = list(product(*[v.values for v in self.vars]))
            probabilitys = []
            for key in product(*[v.values for v in self.vars]):
                probabilitys.append(self.table[indep_key][key])
            
            value = random.choices(values, probabilitys, k=1)[0]
        else:
            values = list(product(*[v.values for v in self.vars]))
            probabilitys = []
            for key in values:
                probabilitys.append(self.table[key])
            
            value = random.choices(values, probabilitys, k=1)[0]
            
        for i,v in enumerate(self.vars):
                v.event = value[i]

File: mib_v2_3_2/distrib.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <optional>
#include "Var.h"

//  Definir una función de hash para std::vector<int>
struct VectorHash {
    std::size_t operator()(const std::vector<int>& v) const {
        std::size_t seed = v.size();
        for (int i : v) {
           // Combinar los hashes de los elementos individuales del vector
           seed ^= std::hash<int>{}(i) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
        }
        return seed;
    }
};



class Distrib {
    private:
        std::unordered_map<int,float> table;
        std::vector<Var> vars;
        std::optional<std::vector<Var>> parents;
        
    public:

    Distrib(const std::unordered_map<int,float>& table, const std::vector<Var>& vars,
        const std::optional<std::vector<Var>>& parents) :
        table(table), vars(vars), parents(parents) {}

    float P() {

    }

};
